### ISSUE TITLE
Why are TCP and gripper latent actions processed separately during inference in real_runner.py?

### ISSUE THREAD

**Issue by colonzyx** [xiaoxiaoxh/reactive_diffusion_policy#19]
Hi, thank you again for your excellent work and detailed codebase!

Apologies for asking yet another question — I truly appreciate your time and patience.

While reviewing the code in reactive_diffusion_policy/env_runner/real_runner.py, I noticed that at lines 358–359, the TCP and gripper latent actions are processed separately using policy.predict_from_latent_action:

full_tcp_step_action = policy.predict_from_latent_action(
    tcp_step_latent_action, extended_obs_dict, tcp_extended_obs_step,
    dataset_obs_temporal_downsample_ratio, extend_obs_pad_after=True
)['action'][0].detach().cpu().numpy()

full_gripper_step_action = policy.predict_from_latent_action(
    gripper_step_latent_action, extended_obs_dict, gripper_extended_obs_step,
    dataset_obs_temporal_downsample_ratio, extend_obs_pad_after=True
)['action'][0].detach().cpu().numpy()

I’m curious about the rationale behind this separation:

Why are TCP and gripper actions handled separately here instead of jointly in a single call?
Since the model is trained with combined TCP and gripper data, does this separation during inference have any effect on consistency or performance?

Thank you again for your support and your amazing contributions!

**Comment by xiaoxiaoxh** [comment]
That is a very good question! In the training dataset, the gripper action and the TCP pose are nearly synchronized. But during inference, the gripper action latency is much higher than the TCP action latency (gripper move slower). If you send an gripper action and a TCP action to the robot at the same time, the gripper will reach that position later than the robot arm. So we set different latency steps to compensate this problem  by discarding the first few steps of the action chunk. Perhaps this figure in the UMI paper will better illustrate this.

<img width="2414" height="925" alt="Image" src="https://github.com/user-attachments/assets/289976fa-8318-4d3b-b8a4-7024f42fc99c" />
